### PR TITLE
Feature/electron storage

### PR DESCRIPTION
**Pull request by arokor**
1. Expose hardware and OS info
2. Add key-value store for storing credentials
3. Add StorageIntegration for the cache

### CODE REVIEW

File: packages/platforms/electron-player/src/main/api/file-system.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        return reject(
+          Error(`Failed to download file, status code: ${response.statusCode}`)
+        );
+      }
+      const writeStream = createWriteStream(destPath);
+      response.on('data', (chunk) => {
+        writeStream.write(chunk);
+      });
+      response.on('end', () => {
+        writeStream.end();
```
Comment (manast) [inline packages/platforms/electron-player/src/main/api/file-system.ts:186]: a handler for error is missing for both request and response.

**Review by manast — COMMENTED** [review]
Looks good! I wrote a few comments though.

File: packages/platforms/electron-player/src/main/api/file-system.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+import { createWriteStream } from 'fs';
+import { join } from 'path';
+import { URL } from 'url';
+import { net } from 'electron';
+
+// Only works with CommonJS require
+const checkDiskSpace = require('check-disk-space').default;
+
+// Base directory for storage
+const BASE_DIR = join(__dirname, 'file-storage');
```
[the remark this answers is not in the document]
Comment (arokor) [reply packages/platforms/electron-player/src/main/api/file-system.ts:18]: A very good point! It might work but is probably more safe to use another path.

File: packages/platforms/electron-player/src/main/api/file-system.ts [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    console.error('Failed to list files:', error);
+    throw error;
+  }
+}
+
+/*
+ * Store a file in the storage. If data is provided, it is used as the file
+ * content. Otherwise, the file is downloaded from the URL.
+ */
+export async function storeFile(
```
[the remark this answers is not in the document]
Comment (arokor) [reply packages/platforms/electron-player/src/main/api/file-system.ts:84]: I think downloading and renaming makes sense. I'll fix that. Regarding the case with stalling downloads, it should timeout at some point so I think that will be managable

File: packages/platforms/electron-player/src/main/api/file-system.ts [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    const stats = await stat(filePath);
+    return {
+      result: { code: StoreResult.Success },
+      item: {
+        url: filePath,
+        size: stats.size,
+      },
+    };
+  } catch (error: any) {
+    console.error('Failed to store file:', error);
```
[the remark this answers is not in the document]
Comment (arokor) [reply packages/platforms/electron-player/src/main/api/file-system.ts:110]: Yes we should

File: packages/platforms/electron-player/src/main/api/file-system.ts (hunk 1 above)
Comment (arokor) [reply packages/platforms/electron-player/src/main/api/file-system.ts:186]: Will fix

File: packages/platforms/electron-player/src/main/api/file-system.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+import { createHash } from 'crypto';
+import { writeFile, mkdir, readdir, unlink, stat } from 'fs/promises';
+import { createWriteStream } from 'fs';
+import { join } from 'path';
+import { URL } from 'url';
+import { net } from 'electron';
+
+// Only works with CommonJS require
+const checkDiskSpace = require('check-disk-space').default;
```
[the remark this answers is not in the document]
Comment (arokor) [reply packages/platforms/electron-player/src/main/api/file-system.ts:15]: I'm getting the following error:
TypeError: checkDiskSpace.checkDiskSpace is not a function

File: packages/platforms/electron-player/src/main/api/file-system.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (arokor) [reply packages/platforms/electron-player/src/main/api/file-system.ts:132]: fixed

File: packages/platforms/electron-player/src/main/api/file-system.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (arokor) [reply packages/platforms/electron-player/src/main/api/file-system.ts:91]: fixed

File: packages/platforms/electron-player/src/main/api/file-system.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (arokor) [reply packages/platforms/electron-player/src/main/api/file-system.ts:16]: I gave this another try and asked ChatGPT for help but this was the only way I could get this to work. Setting esModuleInterop in tsconfig etc didn't work. Other modules work fine to import this way and I think something is wrong in the module itself.

File: packages/platforms/electron-player/src/main/api/file-system.ts (hunk 1 above)
Comment (arokor) [reply packages/platforms/electron-player/src/main/api/file-system.ts:208]: fixed

File: packages/cache/src/storage.integration.ts [hunk 6]
```diff
@@ -32,6 +32,11 @@ export enum StoreError {
   NotEnoughSpace,
 }
 
+export interface StoreFileReturnValue {
```
Comment (manast) [inline packages/cache/src/storage.integration.ts:35]: I have been thinking about this situation that happens quite often where we need to export interfaces. Because normally every interface and class should be in their own file. However, I realize that many times, interfaces like this one that are ONLY needed for the class or interface in the same file, does not make sense to move to a separate file, as that would in fact pollute the file structure and make it less clear for what a given interface is supposed to be used. Thoughts?

File: packages/platforms/electron-player/src/main/api/file-system.ts [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+import { createWriteStream } from 'fs';
+import { join } from 'path';
+import { URL } from 'url';
+import { net } from 'electron';
+import os from 'os';
+
+// Only works with CommonJS require
+const checkDiskSpace = require('check-disk-space').default;
+
+console.log('checkDiskSpace', checkDiskSpace);
```
Comment (manast) [inline packages/platforms/electron-player/src/main/api/file-system.ts:18]: console.log

File: packages/platforms/electron-player/src/main/api/file-system.ts [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+): Promise<StoreFileReturnValue> {
+  // Must be imported dynamically since the main process is a CommonJS module but @castmill/cache is an ES module
+  const { StoreResult } = await import('@castmill/cache');
+
+  const fullPath = join(BASE_DIR, storagePath);
+  try {
+    const filename = getFileName(url);
+
+    const filePath = join(fullPath, filename);
+    const tempPath = filePath + '.tmp';
```
Comment (manast) [inline packages/platforms/electron-player/src/main/api/file-system.ts:104]: This is not important right now, but in the future it would be better to have the tmp directory as a configurable path, as in systems like linux for example there are special directories that get cleaned automatically for this exact purpose.

**Review by manast — APPROVED** [review]
Looks good to me. It would be great with more tests for testing edge cases in the file system integration and so on though.

File: packages/cache/src/storage.integration.ts (hunk 6 above)
Comment (arokor) [reply packages/cache/src/storage.integration.ts:35]: I think that when an interface is used like this, as a type defining an in-parameter or return value of a function it should be defined in the same file as the function itself. Defining it in a separate file just makes the "real" interfaces hard to find among all the return types.

**Review by manast — APPROVED** [review]
